src/main/java/io/vertx/core/net/impl/pool/ConnectionPool.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2011-2021 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+package io.vertx.core.net.impl.pool;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.core.http.ConnectionPoolTooBusyException;
+import io.vertx.core.impl.EventLoopContext;
+import io.vertx.core.net.impl.clientconnection.ConnectResult;
+import io.vertx.core.net.impl.clientconnection.Lease;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+/**
+ * Connection pool.
+ */
+public interface ConnectionPool<C> {
+
+  /**
+   * Acquire a connection from the pool.
+   *
+   * @param context the context
+   * @param weight the weight
+   * @param handler the callback
+   */
+  void acquire(EventLoopContext context, int weight, Handler<AsyncResult<Lease<C>>> handler);
```

Comment:
`EventLoopContext` prevents the cases that maybe `DuplicateContext`, please refer to [pr comment to mail-client](https://github.com/vert-x3/vertx-mail-client/pull/160#issuecomment-768231304) for more detail.

Reply:
can you for now extract the event loop context out of the DuplicateContext with instanceof ?

Reply:
`DuplicateContext` is a package level class.

Reply:
I gave some thoughts to this and in this case I think we need to address this in the mail client.

I believe the mail client should distinguish the context of the TCP connection and the context of the client interacting with the mail asynchronous API. 

Both are separate, specially with pooled connections, e.g a client uses a pooled connection, the connection is created with the context of this client, the connection goes back in the pool, another client comes and reuse this connection and the two contexts might be different. This is the shared pool scenario and this is something that we don't handle well in Vert.x 3 with Vert.x/Netty based connections and we are addressing in Vert.x 4.

The current example of this is the SQL/HTTP client that behave differently in V3/V4, assuming two verticles are using a shared client:
- in V3, a verticle can acquire a connection created by another verticle and get a callback on the other verticle context
- in V4, the callback will happen on the correct context

The solution to this is that the connection and the client API have each their own context and both use message passing to communicate. Of course if the context is the same, then this is a straight execution.

You can check Http1xServerRequest/Http1xConnection to see how it operates. We need I think first to refactor the mail client to behave likewise and then the usage of this pool will be more natural.

Reply:
Thanks for the info, I think it can be addressed in mail client.